File: eglholder/src/main/java/com/aaron/eglholder/EglHelper.java
package com.aaron.eglholder;

import android.opengl.GLDebugHelper;
import android.util.Log;

import javax.microedition.khronos.egl.*;
import javax.microedition.khronos.opengles.GL;
import java.io.Writer;
import java.lang.ref.WeakReference;

import static com.aaron.eglholder.EGLHolder.*;

/**
 * An EGL helper class.
 */

class EglHelper {

    private WeakReference<EGLHolder> mGLSurfaceViewWeakRef;
    EGL10 mEgl;
    EGLDisplay mEglDisplay;
    EGLSurface mEglSurface;
    EGLConfig mEglConfig;
    EGLContext mEglContext;

    public EglHelper(WeakReference<EGLHolder> glSurfaceViewWeakRef) {
        mGLSurfaceViewWeakRef = glSurfaceViewWeakRef;
    }

    public void start() {
        if (LOG_EGL) {
            Log.w("EglHelper", "start() tid=" + Thread.currentThread().getId());
        }
        /*
         * Get an EGL instance
         */
        mEgl = (EGL10) EGLContext.getEGL();

        /*
         * Get to the default display.
         */
        mEglDisplay = mEgl.eglGetDisplay(EGL10.EGL_DEFAULT_DISPLAY);

        if (mEglDisplay == EGL10.EGL_NO_DISPLAY) {
            throw new RuntimeException("eglGetDisplay failed");
        }

        /*
         * We can now initialize EGL for that display
         */
        int[] version = new int[2];
        if (!mEgl.eglInitialize(mEglDisplay, version)) {
            throw new RuntimeException("eglInitialize failed");
        }
        EGLHolder view = mGLSurfaceViewWeakRef.get();
        if (view == null) {
            mEglConfig = null;
            mEglContext = null;
        } else {
            mEglConfig = view.mEGLConfigChooser.chooseConfig(mEgl, mEglDisplay);

            /*
             * Create an EGL context. We want to do this as rarely as we can, because an
             * EGL context is a somewhat heavy object.
             */
            mEglContext = view.mEGLContextFactory.createContext(mEgl, mEglDisplay, mEglConfig);
        }
        if (mEglContext == null || mEglContext == EGL10.EGL_NO_CONTEXT) {
            mEglContext = null;
            throwEglException("createContext");
        }
        if (LOG_EGL) {
            Log.w("EglHelper", "createContext " + mEglContext + " tid=" + Thread.currentThread().getId());
        }

        mEglSurface = null;
    }

    /**
     * Create an egl surface for the current SurfaceHolder surface. If a surface
     * already exists, destroy it before creating the new surface.
     *
     * @return true if the surface was created successfully.
     */
    public boolean createSurface() {
        if (LOG_EGL) {
            Log.w("EglHelper", "createSurface()  tid=" + Thread.currentThread().getId());
        }
        /*
         * Check preconditions.
         */
        if (mEgl == null) {
            throw new RuntimeException("egl not initialized");
        }
        if (mEglDisplay == null) {
            throw new RuntimeException("eglDisplay not initialized");
        }
        if (mEglConfig == null) {
            throw new RuntimeException("mEglConfig not initialized");
        }

        /*
         *  The window size has changed, so we need to create a new
         *  surface.
         */
        destroySurfaceImp();

        /*
         * Create an EGL surface we can render into.
         */
        EGLHolder view = mGLSurfaceViewWeakRef.get();
        if (view != null) {
            mEglSurface = view.mEGLWindowSurfaceFactory.createWindowSurface(mEgl,
                    mEglDisplay, mEglConfig, view.surface);
        } else {
            mEglSurface = null;
        }

        if (mEglSurface == null || mEglSurface == EGL10.EGL_NO_SURFACE) {
            int error = mEgl.eglGetError();
            if (error == EGL10.EGL_BAD_NATIVE_WINDOW) {
                Log.e("EglHelper", "createWindowSurface returned EGL_BAD_NATIVE_WINDOW.");
            }
            return false;
        }

        /*
         * Before we can issue GL commands, we need to make sure
         * the context is current and bound to a surface.
         */
        if (!mEgl.eglMakeCurrent(mEglDisplay, mEglSurface, mEglSurface, mEglContext)) {
            /*
             * Could not make the context current, probably because the underlying
             * SurfaceView surface has been destroyed.
             */
            logEglErrorAsWarning("EGLHelper", "eglMakeCurrent", mEgl.eglGetError());
            return false;
        }

        return true;
    }

    /**
     * Create a GL object for the current EGL context.
     *
     * @return
     */
    GL createGL() {

        GL gl = mEglContext.getGL();
        EGLHolder view = mGLSurfaceViewWeakRef.get();
        if (view != null) {
            if (view.mGLWrapper != null) {
                gl = view.mGLWrapper.wrap(gl);
            }

            if ((view.mDebugFlags & (DEBUG_CHECK_GL_ERROR | DEBUG_LOG_GL_CALLS)) != 0) {
                int configFlags = 0;
                Writer log = null;
                if ((view.mDebugFlags & DEBUG_CHECK_GL_ERROR) != 0) {
                    configFlags |= GLDebugHelper.CONFIG_CHECK_GL_ERROR;
                }
                if ((view.mDebugFlags & DEBUG_LOG_GL_CALLS) != 0) {
                    log = new LogWriter();
                }
                gl = GLDebugHelper.wrap(gl, configFlags, log);
            }
        }
        return gl;
    }

    /**
     * Display the current render surface.
     *
     * @return the EGL error code from eglSwapBuffers.
     */
    public int swap() {
        if (!mEgl.eglSwapBuffers(mEglDisplay, mEglSurface)) {
            return mEgl.eglGetError();
        }
        return EGL10.EGL_SUCCESS;
    }

    public void destroySurface() {
        if (LOG_EGL) {
            Log.w("EglHelper", "destroySurface()  tid=" + Thread.currentThread().getId());
        }
        destroySurfaceImp();
    }

    private void destroySurfaceImp() {
        if (mEglSurface != null && mEglSurface != EGL10.EGL_NO_SURFACE) {
            mEgl.eglMakeCurrent(mEglDisplay, EGL10.EGL_NO_SURFACE,
                    EGL10.EGL_NO_SURFACE,
                    EGL10.EGL_NO_CONTEXT);
            EGLHolder view = mGLSurfaceViewWeakRef.get();
            if (view != null) {
                view.mEGLWindowSurfaceFactory.destroySurface(mEgl, mEglDisplay, mEglSurface);
            }
            mEglSurface = null;
        }
    }

    public void finish() {
        if (LOG_EGL) {
            Log.w("EglHelper", "finish() tid=" + Thread.currentThread().getId());
        }
        if (mEglContext != null) {
            EGLHolder view = mGLSurfaceViewWeakRef.get();
            if (view != null) {
                view.mEGLContextFactory.destroyContext(mEgl, mEglDisplay, mEglContext);
            }
            mEglContext = null;
        }
        if (mEglDisplay != null) {
            mEgl.eglTerminate(mEglDisplay);
            mEglDisplay = null;
        }
    }

    private void throwEglException(String function) {
        throwEglException(function, mEgl.eglGetError());
    }

    public static void throwEglException(String function, int error) {
        String message = formatEglError(function, error);
        if (LOG_THREADS) {
            Log.e("EglHelper", "throwEglException tid=" + Thread.currentThread().getId() + " "
                    + message);
        }
        throw new RuntimeException(message);
    }

    public static void logEglErrorAsWarning(String tag, String function, int error) {
        Log.w(tag, formatEglError(function, error));
    }

    public static String formatEglError(String function, int error) {
        return function + " failed: " + getErrorString(error);
    }

    public static String getErrorString(int error) {
        switch (error) {
            case EGL11.EGL_SUCCESS:
                return "EGL_SUCCESS";
            case EGL11.EGL_NOT_INITIALIZED:
                return "EGL_NOT_INITIALIZED";
            case EGL11.EGL_BAD_ACCESS:
                return "EGL_BAD_ACCESS";
            case EGL11.EGL_BAD_ALLOC:
                return "EGL_BAD_ALLOC";
            case EGL11.EGL_BAD_ATTRIBUTE:
                return "EGL_BAD_ATTRIBUTE";
            case EGL11.EGL_BAD_CONFIG:
                return "EGL_BAD_CONFIG";
            case EGL11.EGL_BAD_CONTEXT:
                return "EGL_BAD_CONTEXT";
            case EGL11.EGL_BAD_CURRENT_SURFACE:
                return "EGL_BAD_CURRENT_SURFACE";
            case EGL11.EGL_BAD_DISPLAY:
                return "EGL_BAD_DISPLAY";
            case EGL11.EGL_BAD_MATCH:
                return "EGL_BAD_MATCH";
            case EGL11.EGL_BAD_NATIVE_PIXMAP:
                return "EGL_BAD_NATIVE_PIXMAP";
            case EGL11.EGL_BAD_NATIVE_WINDOW:
                return "EGL_BAD_NATIVE_WINDOW";
            case EGL11.EGL_BAD_PARAMETER:
                return "EGL_BAD_PARAMETER";
            case EGL11.EGL_BAD_SURFACE:
                return "EGL_BAD_SURFACE";
            case EGL11.EGL_CONTEXT_LOST:
                return "EGL_CONTEXT_LOST";
            default:
                return getHex(error);
        }
    }

    private static String getHex(int value) {
        return "0x" + Integer.toHexString(value);
    }

}

File: egllibrary/src/main/java/com/aaron/egl/EglHelper.kt
package com.aaron.egl

import android.content.ContentValues.TAG
import android.opengl.*
import android.opengl.EGL14.*
import android.util.Log

internal class EglHelper {
//    var mEgl: EGL10? = null
    var mEglDisplay: EGLDisplay? = null
    var mEglSurface: EGLSurface? = null
    var mEglConfig: EGLConfig? = null
    var mEglContext: EGLContext? = null
    var mEGLContextClientVersion: Int = 2

    var mEGLConfigChooser: EGLConfigChooser? = null
        set(value) {
            if (checkRenderThreadState()) {
                field = value
            }
        }
    var mEGLContextFactory: EGLContextFactory? = null
        set(value) {
            if (checkRenderThreadState()) {
                field = value
            }
        }
    var mEGLWindowSurfaceFactory: EGLWindowSurfaceFactory? = null
        set(value) {
            if (checkRenderThreadState()) {
                field = value
            }
        }

    fun createEglContext() : Boolean {
        if (mEGLConfigChooser == null) {
            mEGLConfigChooser = RecordableEGLConfigChooser(mEGLContextClientVersion)
        }
        if (mEGLContextFactory == null) {
            mEGLContextFactory = DefaultContextFactory(mEGLContextClientVersion)
        }
        if (mEGLWindowSurfaceFactory == null) {
            mEGLWindowSurfaceFactory = DefaultWindowSurfaceFactory()
        }

//        mEgl = EGLContext.getEGL() as EGL10
        mEglDisplay = eglGetDisplay(EGL_DEFAULT_DISPLAY)

        if (mEglDisplay === EGL_NO_DISPLAY) {
            Log.e(TAG, "eglGetDisplay failed")
            destroyEglContext()
            return false
        }
        val version = IntArray(2)
        if (!eglInitialize(mEglDisplay, version, 0, version, 1)) {
            Log.e(TAG, "eglInitialize failed")
            destroyEglContext()
            return false
        }
        mEglConfig = mEGLConfigChooser!!.chooseConfig(mEglDisplay)

        /*
         * Create an EGL context. We want to do this as rarely as we can, because an
         * EGL context is a somewhat heavy object.
         */
        mEglContext = mEGLContextFactory!!.createContext(mEglDisplay, mEglConfig)
        if (mEglContext === EGL_NO_CONTEXT) {
            Log.e(TAG, "createContext")
            destroyEglContext()
            return false
        }
        mEglSurface = null
        return true
    }

    /**
     * Create an egl surface for the current SurfaceHolder surface. If a surface
     * already exists, unregister it before creating the new surface.
     *
     * @return true if the surface was created successfully.
     */
    fun createSurface(displayObject: Any): Boolean {
        /*
         * Check preconditions.
         */
        if (mEglDisplay == null) {
            Log.e(TAG, "createSurface eglDisplay not initialized")
            return false
        }
        if (mEglConfig == null) {
            Log.e(TAG, "createSurface mEglConfig not initialized")
            return false
        }

        /*
         *  The window size has changed, so we need to create a new
         *  surface.
         */
        destroySurfaceImp()
        mEGLWindowSurfaceFactory!!.setNativeWindow(displayObject)
        /*
         * Create an EGL surface we can render into.
         */
        mEglSurface = mEGLWindowSurfaceFactory!!.createWindowSurface( mEglDisplay, mEglConfig)

        if (mEglSurface == null || mEglSurface === EGL_NO_SURFACE) {
            val error = eglGetError()
            if (error == EGL_BAD_NATIVE_WINDOW) {
                Log.e(TAG, "createSurface false : EGL_BAD_NATIVE_WINDOW.")
            } else {
                Log.e(TAG, "createSurface false .")
            }
            destroySurfaceImp()
            return false
        }

        /*
         * Before we can issue GL commands, we need to make sure
         * the context is current and bound to a surface.
         */
        if (!EGLUtils.makeCurrentSurface(mEglDisplay, mEglContext, mEglSurface)) {
            /*
             * Could not make the context current, probably because the underlying
             * SurfaceView surface has been destroyed.
             */
            logEglErrorAsWarning("EGLHelper", "eglMakeCurrent", eglGetError())
            return false
        }

        return true
    }


    /**
     * Display the current render surface.
     *
     * @return the EGL error code from eglSwapBuffers.
     */
    fun swap(): Int {
        return if (!EGLUtils.swapSurface(mEglDisplay, mEglSurface)) {
            eglGetError()
        } else EGL_SUCCESS
    }

    fun destroySurface() {
        destroySurfaceImp()
    }

    private fun destroySurfaceImp() {
        if (mEglSurface != null && mEglSurface !== EGL_NO_SURFACE) {
            eglMakeCurrent(
                mEglDisplay, EGL_NO_SURFACE,
                EGL_NO_SURFACE,
                EGL_NO_CONTEXT
            )
            mEGLWindowSurfaceFactory!!.destroySurface(mEglDisplay, mEglSurface)
            mEglSurface = null
        }
    }

    fun destroyEglContext() {
        if (mEglContext != null) {
            mEGLContextFactory!!.destroyContext(mEglDisplay, mEglContext)
            mEglContext = null
        }
        if (mEglDisplay != null) {
            eglTerminate(mEglDisplay)
            mEglDisplay = null
        }
    }

    private fun throwEglException(function: String) {
        throwEglException(function, eglGetError())
    }

    private fun checkRenderThreadState(): Boolean {
        if (mEglContext != null) {
            return false
        }
        return true
    }

    companion object {

        fun throwEglException(function: String, error: Int) {
            val message = formatEglError(function, error)
//            Logger.w(
//                "EglHelper", "throwEglException tid=" + Thread.currentThread().id + " "
//                        + message
//            )
            throw RuntimeException(message)
        }

        fun logEglErrorAsWarning(tag: String, function: String, error: Int) {
//            Logger.w(tag, formatEglError(function, error))
        }

        fun formatEglError(function: String, error: Int): String {
            return "$function failed: $error"
        }
    }
}
